### PR TITLE
bpo-34113: Fix SIGSEGV on negative STACKADJ when LLTRACE is on

### PR DESCRIPTION
**Pull request by costypetrisor**
https://bugs.python.org/issue34113

https://bugs.python.org/issue34113

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


When your account is ready, please add a comment in this pull request
and a Python core developer will remove the `CLA not signed` label
to make the bot check again.

You can [check yourself](https://devguide.python.org/committing.html#checking-if-the-cla-has-been-received)
to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by costypetrisor** [comment]
cc @vstinner

**Comment by encukou** [comment]
Can you also add a test that fails without this change?

**Comment by costypetrisor** [comment]
@encukou I pushed a new test which without the fix would crash

File: Lib/test/test_lltrace.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+
```
Comment (vstinner) [inline Lib/test/test_lltrace.py:4]: please remove this empty line

File: Lib/test/test_lltrace.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import os
+import tempfile
+import textwrap
+import unittest
+
+from test import support
+from test.support.script_helper import (
+    assert_python_ok, assert_python_failure, interpreter_requires_environment,
```
Comment (vstinner) [inline Lib/test/test_lltrace.py]: interpreter_requires_environment and assert_python_failure import are not used. please remove unused imports.

File: Lib/test/test_lltrace.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            console = code.InteractiveConsole()
+            console.push('__ltrace__ = 1')
+            console.push('a = [1, 2, 3]')
+            console.push('a[0] = 1')
+            print('unreachable if bug exists')
+            """))
+            tmp_script.flush()
+
+            status, stdout, stderr = assert_python_ok(tmp_script.name)
+            self.assertEqual(status, 0)
```
Comment (vstinner) [inline Lib/test/test_lltrace.py]: the function already checks status. in fact, you don't need the function result: "status, stdout, stderr = " can be removed IMHO.

File: Lib/test/test_lltrace.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class TestLLTrace(unittest.TestCase):
+
+    def test_lltrace_does_not_crash_on_subscript_operator(self):
+        # If this test fails, it will reproduce a crash reported as
+        # bpo-34113. The crash happened at the command line console of
+        # debug Python builds with __ltrace__ enabled (only possible in console),
+        # when the interal Python stack was negatively adjusted
+        with tempfile.NamedTemporaryFile(mode='w+', delete=False) as tmp_script:
```
Comment (vstinner) [inline Lib/test/test_lltrace.py]: I don't think that it's a good practice to keep the file open while running the script in a subprocess.

I suggest to use support.TESTFN filename, and use self.addCleanup(support.remove, support.TESTFN) before creating the file.

File: Lib/test/test_lltrace.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            console.push('a[0] = 1')
+            print('unreachable if bug exists')
+            """))
+            tmp_script.flush()
+
+            status, stdout, stderr = assert_python_ok(tmp_script.name)
+            self.assertEqual(status, 0)
+
+
+def test_main():
```
Comment (vstinner) [inline Lib/test/test_lltrace.py]: This function is not needed, it can be removed.

File: Lib/test/test_lltrace.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    support.run_unittest(
+        TestLLTrace,
+    )
+
+if __name__ == "__main__":
+    test_main()
```
Comment (vstinner) [inline Lib/test/test_lltrace.py]: Calling unittest.main() here is just fine. Hey, it's Python 3, unittest now has cool features!

File: Python/ceval.c [hunk 7]
```diff
@@ -762,16 +762,26 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
...
-                          assert(STACK_LEVEL() <= co->co_stacksize); }
+#define STACKADJ_GROW(n)    do { \
+                              assert(n >= 0); \
+                              (void)(BASIC_STACKADJ(n), \
+                              lltrace && prtrace(TOP(), "stackadj")); \
+                              assert(STACK_LEVEL() <= co->co_stacksize); \
+                            } while (0)
+#define STACKADJ_SHRINK(n)  do { \
+                              assert(n >= 0); \
+                              (void)(lltrace && prtrace(TOP(), "stackadj")); \
```
Comment (vstinner) [inline Python/ceval.c]: This syntax is strange. Why not using a regular if?

```
if (lltrace) prtrace(TOP(), "stackadj");
```

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Python/ceval.c [hunk 8]
```diff
@@ -762,16 +762,26 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
                           assert(STACK_LEVEL() <= co->co_stacksize); }
 #define POP()           ((void)(lltrace && prtrace(TOP(), "pop")), \
                          BASIC_POP())
-#define STACKADJ(n)     { (void)(BASIC_STACKADJ(n), \
-                          lltrace && prtrace(TOP(), "stackadj")); \
-                          assert(STACK_LEVEL() <= co->co_stacksize); }
+#define STACKADJ_GROW(n)    do { \
```
Comment (vstinner) [inline Python/ceval.c]: Maybe rename methods to: STACK_GROW() and STACK_SHRINK()?

File: Lib/test/test_lltrace.py (hunk 1 above)
Comment (costypetrisor) [reply Lib/test/test_lltrace.py:4]: Done :)

File: Python/ceval.c (hunk 7 above)
Comment (costypetrisor) [reply Python/ceval.c]: I did it like that in order to maintain the the same style as the original author.

I did a quick diff and generate pretty much the same assembly with `-O0 -g` (lltrace is available only on debug builds): https://godbolt.org/g/MMnyrq

Let me know if you prefer a plain if and I'll quickly change

File: Lib/test/test_lltrace.py (hunk 6 above)
Comment (costypetrisor) [reply Lib/test/test_lltrace.py]: Done :)
Love the way you advertise the new stuff :)

**Comment by costypetrisor** [comment]
I have made the requested changes; please review again
:)

**Review by vstinner — COMMENTED** [review]
Hum. I'm confused. Why do you write "LLTRACE" with 2 L whereas the constant only contains 1 L? If it's a typo, would you mind to rename the test file, test case and test method?

**Comment by vstinner** [comment]
Oh, LLTRACE comes from a C define:

```
#define LLTRACE  1      /* Low-level trace feature */
```

LL as Low-Level, I see. It's just surprising that the constant is called ltrace with a single L :-)

**Comment by costypetrisor** [comment]
@vstinner I was confused about this during the sprint too. Petr Viktorin who was sitting next to me didn't knew why this was written like this either.

So from what I understand the gist is that LLTRACE is the feature coming from Low Level Trace, but the global that you set in the console interpretor is `__ltrace__` with just 1 lower-case L.

**Pull request merged**